torchtune/models/llava_next/image_processor.py

```diff
@@ -0,0 +1,344 @@
+        """
+
+        self.possible_resolutions = torch.tensor(possible_resolutions)
+
+        if isinstance(resample, torchvision.transforms.InterpolationMode):
+            self.resample = resample
+        if resample == "bilinear":
+            self.resample = torchvision.transforms.InterpolationMode.BILINEAR
+        elif resample == "bicubic":
+            self.resample = torchvision.transforms.InterpolationMode.BICUBIC
+        else:
+            raise ValueError(
+                "resample must be of type torchvision.transforms.InterpolationMode or ['bilinear', 'bicubic]."
+            )
+
+        self.center_crop = v2.CenterCrop(size=patch_size)
+
+        self.patch_size = patch_size
+
+    @staticmethod
+    def _get_new_size_without_distortion(
+        image_size: Tuple[int, int],
+        target_resolution: List[int],
+    ) -> Tuple[int, int]:
+
+        """
+        Determines the maximum dimensions to which an image can be resized without distorting its
+        aspect ratio, based on the target resolution.
+
+        Args:
+            image_size (Tuple[int, int]): The original dimensions of the image (height, width).
+            target_resolution (List[int]): The desired resolution to fit the image into (height, width).
+        Returns:
+            Tuple[int, int]: The optimal dimensions (height, width) to which the image should be resized.
+        Example:
+            >>> _get_new_size_without_distortion([200, 300], target_resolution = [450, 200])
+            (134, 200)
+            >>> _get_new_size_without_distortion([800, 600], target_resolution = [450, 1300])
+            (450, 338)
+        """
+
+        original_height, original_width = image_size
+        target_height, target_width = target_resolution
+
+        scale_w = target_width / original_width
+        scale_h = target_height / original_height
+
+        if scale_w < scale_h:
+            new_width = target_width
+            new_height = min(math.ceil(original_height * scale_w), target_height)
+        else:
+            new_height = target_height
+            new_width = min(math.ceil(original_width * scale_h), target_width)
+
+        return new_height, new_width
+
+    @staticmethod
+    def _divide_to_patches(image: torch.Tensor, patch_size: int) -> List[torch.Tensor]:
+        """
+        Divides an image into patches. The patches will not have equal size if the
+        aspect ratio of the image is not divisible by the patch size.
+
+        Args:
+            image (torch.Tensor): Image tensor.
+            patch_size (int): Size of each patch.
+        Returns:
+            List[torch.Tensor]: List of image patches.
+        Example:
+            >>> image = torch.rand(3, 200, 300)
+            >>> patches = _divide_to_patches(image, patch_size=50)
+            >>> len(patches)
+            24
+            >>> image = torch.rand(3, 400, 500)
+            >>> patches = _divide_to_patches(image, patch_size=450)
```

Comment:
why does this return 2 patches if 450 > 400? maybe it would be helpful to also print the patch dims? Also what happens if the image dim does not divide evenly by patch size, do we just drop the remainder? We don't resize it?

Reply:
we resize it in later steps. The patches have unequal sizes. I tried to express it in the docstring, but I agree that adding it to the examples would help.